packages/core/src/sync-realtime/service.ts

```diff
@@ -486,160 +374,393 @@ export class RealtimeSyncService extends Emittery<RealtimeSyncEvents> {
+      fromBlock: numberToHex(latestLocalBlockNumber + 1),
+      toBlock: numberToHex(newBlockNumber),
+    });
 
-      const missingBlocks = await Promise.all(missingBlockRequests);
+    const matchedLogs = await this.getMatchedLogs(
+      logs,
+      BigInt(newBlockNumber),
+      true,
+    );
 
-      // Add blocks to the queue from oldest to newest. Include the current block.
-      for (const block of [...missingBlocks, newBlockWithTransactions]) {
-        const priority = Number.MAX_SAFE_INTEGER - hexToNumber(block.number);
-        this.queue.addTask({ block }, { priority });
-      }
+    return { blocks: newBlocks, logs: matchedLogs };
+  };
 
-      this.common.logger.debug({
-        service: "realtime",
-        msg: `Fetched missing blocks [${missingBlockNumbers[0]}, ${
-          missingBlockNumbers[missingBlockNumbers.length - 1]
-        }] (network=${this.network.name})`,
-      });
+  private syncBatch = async (
+    newBlock: RealtimeBlock,
+  ): Promise<{ blocks: RealtimeBlock[]; logs: RealtimeLog[] }> => {
+    const latestLocalBlock = this.getLatestLocalBlock();
+    const latestLocalBlockNumber = latestLocalBlock.number;
 
-      return;
-    }
+    const newBlockNumber = hexToNumber(newBlock.number);
 
-    // 4) There has been a reorg, because:
-    //   a) newBlock.number <= headBlock + 1.
-    //   b) newBlock.hash is not found in our local chain.
-    // which means newBlock is on a fork of our local chain.
-    //
-    // To reconcile, traverse up the remote (canonical) chain until we find the first
-    // block that is present in both chains (the common ancestor block).
+    // Get logs
+    const logs = await this._eth_getLogs({
+      fromBlock: numberToHex(latestLocalBlockNumber + 1),
+      toBlock: newBlock.number,
+    });
 
-    // Store the block objects as we fetch them.
-    // Once we find the common ancestor, we will add these blocks to the queue.
-    const canonicalBlocksWithTransactions = [newBlockWithTransactions];
+    const matchedLogs = await this.getMatchedLogs(
+      logs,
+      BigInt(newBlockNumber),
+      true,
+    );
 
-    // Keep track of the current canonical block
-    let canonicalBlock = newBlock;
-    let depth = 0;
+    // Get blocks
+    const missingBlockNumbers = dedupe(
+      matchedLogs.map((log) => log.blockNumber!),
+    ).filter((b) => b !== newBlock.number);
 
-    this.common.logger.warn({
-      service: "realtime",
-      msg: `Detected reorg with forked block (${canonicalBlock.number}, ${canonicalBlock.hash}) (network=${this.network.name})`,
+    const blocks = await Promise.all(
+      missingBlockNumbers.map(this._eth_getBlockByNumber),
+    );
+    blocks.push(newBlock);
+
+    return { blocks: blocks, logs: matchedLogs };
+  };
+
+  /**
+   * Check if a re-org occurred by comparing remote logs to local.
+   *
+   * @returns True if a re-org has occurred.
+   */
+  detectReorg = async (latestBlockNumber: number) => {
```

Comment:
Still makes sense to me that this would be called `reconcileReorg` or something. The current function signature kinda suggests to me that it's a pure function.